clustering/jgroups/WFLY-9563_Integrate_KUBE_PING_Discovery_Protocol.adoc

```diff
@@ -0,0 +1,92 @@
+= Title
+:author:            Radoslav Husar
+:email:             rhusar@redhat.com
+:toc:               left
+:icons:             font
+:keywords:          openshift,kubernetes,jgroups,clustering
+:idprefix:
+:idseparator:       -
+
+== Overview
+
+`KUBE_PING` is a discovery protocol for JGroups cluster nodes managed by Kubernetes.
+Since Kubernetes is in charge of launching nodes, it knows the IP addresses of all pods it started,
+and is therefore the best place to ask for cluster discovery.
+Discovery is done by asking Kubernetes for a list of IP addresses of all cluster nodes.
+
+Given OpenShift is based on Kubernetes `KUBE_PING`, it is an ideal fit.
+
+A separate integration is needed besides just JGroups upgrade.
+Historically all protocols were bundled within main `jgroups-<version>.jar`.
+Since the main JGroups project does not allow runtime dependencies, protocols had to include a lot of unmaintainable copy-pasted and trimmed library code.
+This hinders supportability.
+Protocols needing to fetch dependencies now reside in separate repository in the https://github.com/jgroups-extras repository.
+As such, `KUBE_PING` resides in https://github.com/jgroups-extras/jgroups-kubernetes repository.
+
+Note that the OpenShift team currently maintain their own implementation of PING protocols to be used in OpenShift.
+See the https://github.com/jboss-openshift/openshift-ping repository.
```

Comment:
This is interesting, I didn't know about that repository. Which `KUBE_PING` protocol are you going to integrate then? I'm thinking that maybe we could avoid some duplication of effort if we use the "already supported" OpenShift fork. Assuming it's a viable choice.

Reply:
The discussions are ongoing, but openshift is just one of many Kubernetes-based deployment so having Kubernetes-based ping protocol is generally useful. Unfortunately, it's not possible to use 'their' version, since its not publicly released into maven repositories. Also, IIRC they currently use DNS ping also from their own repo (not the jgroups-managed version).